torax/sources/electron_cyclotron_source.py

```diff
@@ -0,0 +1,197 @@
+    # - Flat-top plasma operational space of the STEP power plant
+    #   Tholerus et al., 2024
+    #   https://doi.org/10.1088/1741-4326/ad6ea2
+    #
+    # Note an additional B0 term appears, as TORAX expects <j.B> rather than <j.B>/B0
+    #
+    # Units:
+    # - epsilon0^2: [m^-6 kg^-2 s^8 A^4]
+    # - qe^-3: [C^-3] = [A^-3 s^-3]
+    # - global_efficiency: [dimensionless]
+    # - ec_power_density: [W/m^3] = [kg m^-1 s^-3]
+    # - Te: [J] = [kg m^2 s^-2]
+    # - ne^-1: [m^3]
+
+    total_ec_power = jax.scipy.integrate.trapezoid(
+        dynamic_source_runtime_params.ec_power_density * geo.vpr_face, geo.rho_face_norm
+    )
+    weighted_ec_power = jax.scipy.integrate.trapezoid(
+        dynamic_source_runtime_params.ec_power_density
+        * core_profiles.temp_el.face_value()
+        / core_profiles.ne.face_value()
+        * geo.vpr_face,
+        geo.rho_face_norm,
+    )
+    local_efficiency = weighted_ec_power / total_ec_power
+    # Compute via the log for numerical stability
+    # This is equivalent to:
+    # j_ec_parallel = (
+    #     global_efficiency
+    #     * 2 * jnp.pi * epsilon0 ** 2
+    #     / qe ** -3
+    #     * local_efficiency (in J and m^-3)
+    #     * ec_power_density
+    # )
+    log_j_ec_parallel = (
+        jnp.log(dynamic_source_runtime_params.global_efficiency)
+        + jnp.log(2 * jnp.pi)
+        + 2 * jnp.log(CONSTANTS.epsilon0)
+        - 3 * jnp.log(CONSTANTS.qe)
+        + jnp.log(local_efficiency)
+        # Convert from keV to J - moved inside the log from the local_efficiency integral
+        + jnp.log(CONSTANTS.keV2J)
+        # Convert from nref to m^-3 - moved inside the log from the local_efficiency integral
+        - jnp.log(dynamic_runtime_params_slice.numerics.nref)
+        + jnp.log(dynamic_source_runtime_params.ec_power_density)
+    )
+    j_ec_parallel = jnp.exp(log_j_ec_parallel)
+
+    return jnp.stack(
+        [dynamic_source_runtime_params.ec_power_density, j_ec_parallel * geo.B0]
+    )
+
+
+def _get_ec_output_shape(geo: geometry.Geometry) -> tuple[int, ...]:
+    return (2,) + source.ProfileType.CELL.get_profile_shape(geo)
+
+
+@dataclasses.dataclass(kw_only=True, frozen=True, eq=True)
+class ElectronCyclotronSource(source.Source):
+    """Electron cyclotron source for the Psi and Te equations."""
+
+    affected_core_profiles: tuple[source.AffectedCoreProfile, ...] = dataclasses.field(
+        init=False,
+        default=(
+            source.AffectedCoreProfile.TEMP_EL,
+            source.AffectedCoreProfile.PSI,
+        ),
+    )
+
+    output_shape_getter: source.SourceOutputShapeFunction = dataclasses.field(
+        init=False,
+        default_factory=lambda: _get_ec_output_shape,
+    )
+
+    supported_modes: tuple[runtime_params_lib.Mode, ...] = (
+        runtime_params_lib.Mode.ZERO,
+        runtime_params_lib.Mode.MODEL_BASED,
+    )
+
+    model_func: source.SourceProfileFunction = _calc_heating_and_linliu_current
+
+    def get_value(
```

Comment:
Can `get_value` be removed, and purely inherited from `source.Source`? This pattern was copied over from `IonElectronSource`, and I'm not sure whether that is up to date / the best way to be doing things.

Reply:
I believe it can yes :)